### ISSUE TITLE
WebhookEvent.get_resource() raises KeyError

### ISSUE THREAD

**Issue by wolfadactyl** [paypal/PayPal-Python-SDK#173]
### General Information

- Mode (Sandbox/Live): sandbox
- Version of Python used: 3.5.1
- SDK Version: 1.12.0

### Issue Description

When using the [webhooks simulator](https://developer.paypal.com/developer/webhooksSimulator) to send events, loading the request data into a WebhookEvent and calling get_resource results in a KeyError for the given resource_type. So far I've tried events for authorization and payouts-related events and they all have the same issue.

From what I can see it looks like there are different resource_types being sent than what is expected. WebhookEvent.convert_resources contains keys for `authorization`, `payout`, and [`payoutItem`, `payoutitem`, `payout-item`] but the events are sending resource_types of `authorization_consent_revoked`, `payouts`, and `payouts_item`. I'm not sure whether or not this is also happening on live, or if the simulator is incorrect, but there most likely needs to be a pass to align them all accordingly.

**Comment by jaypatel512** [comment]
Hey @wolfadactyl !

The webhooks simulator is only used to test your integration receiving the webhooks. The data returned in the webhook event is hard coded, and static. The `authorization` resource you receive back in that webhook is not really a valid authorization, and will not work for any lookup calls, as well as other operations.

To test the entire flow, we would recommend using a real webhook event, and get webhook event to your local machine using  `ngrok` tool. This should allow you to debug webhook integration in your local machine.

Please re-open if that does not fix the issue.

**Comment by wolfadactyl** [comment]
@jaypatel512 apparently I can't re-open this myself =(

**Comment by jaypatel512** [comment]
Re-opened.

There will be some naming convention differences between the resource type, and object type defined in Python. The issue is, every SDK (language) have different naming conventions, and its hard to get all languages to conform with same naming style. 

We have this method https://github.com/paypal/PayPal-Python-SDK/blob/master/paypalrestsdk/util.py#L58 that figures out the resource type, and fetch the resource. Most likely, it needs to have a mapping table to catch these cases.

**Comment by wolfadactyl** [comment]
Yeah I tried forking and adding `payouts` and `payouts_item` to the respective convert_resources spots but that didn't appear to work:
```
Payout.convert_resources['payouts'] = Payout
PayoutItem.convert_resources['payouts_item'] = PayoutItem
```
Given that the paths for each resource are `payouts` and `payouts-item`, though, I would think this change indicates an alignment between the path and the resource_type, no? I'd be surprised if any of the other SDKs are handling this properly if they're expecting a singularly named resource as well.

**Comment by someidiot** [comment]
No updates on this?
@wolfadactyl did you get anything working?

**Comment by tkrevh** [comment]
Same problem here. SDK can not parse the payload received from the sandbox simulator.
Error: **KeyError: u'payouts_item'**

Payload:
`{
    "id": "WH-68L35871MU875914X-8DM79009TB591331B",
    "event_version": "1.0",
    "create_time": "2016-06-06T03:44:51Z",
    "resource_type": "payouts_item",
    "event_type": "PAYMENT.PAYOUTS-ITEM.SUCCEEDED",
    "summary": "A payout item has succeeded",
    "resource": {
        "payout_item_id": "AYNYWNCHBD8KS",
        "transaction_id": "57J64166G9424913F",
        "transaction_status": "SUCCESS",
        "payout_item_fee": {
            "currency": "USD",
            "value": "0.25"
        },
        "payout_batch_id": "Q8B9WFS7ZZJ4Q",
        "payout_item": {
            "recipient_type": "EMAIL",
            "amount": {
                "currency": "USD",
                "value": "1.0"
            },
            "note": "First payout",
            "receiver": "beamdaddy@paypal.com",
            "sender_item_id": "Item1"
        },
        "time_processed": "2016-06-06T03:44:51Z",
        "links": [
            {
                "href": "https://api.paypal.com/v1/payments/payouts-item/AYNYWNCHBD8KS",
                "rel": "self",
                "method": "GET"
            },
            {
                "href": "https://api.paypal.com/v1/payments/payouts/Q8B9WFS7ZZJ4Q",
                "rel": "batch",
                "method": "GET"
            }
        ]
    },
    "links": [
        {
            "href": "https://api.paypal.com/v1/notifications/webhooks-events/WH-68L35871MU875914X-8DM79009TB591331B",
            "rel": "self",
            "method": "GET"
        },
        {
            "href": "https://api.paypal.com/v1/notifications/webhooks-events/WH-68L35871MU875914X-8DM79009TB591331B/resend",
            "rel": "resend",
            "method": "POST"
        }
    ]
}`

Steps to reproduce (take the payload above):
```
from paypalrestsdk.notifications import WebhookEvent
webhook_event = WebhookEvent(payload)
resource = webhook_event .get_resource()
```

**Comment by tkrevh** [comment]
I can get it to work by doing a modification to the following method:
[](https://github.com/paypal/PayPal-Python-SDK/blob/master/paypalrestsdk/util.py#L58)

```
def get_member(name):
    """
    Get the paypalrestsdk member class represented by name. Helper
    method for fetching resource sent via webhook event

    Usage::

    >>> util.get_member('authorization')
    <class 'paypalrestsdk.payments.Authorization'>
    """
    mapping_table = {
        'payouts_item': 'paypalrestsdk.payments.PayoutItem'
    }
    resource_class_dict = dict((k.lower(), ("{0}.{1}".format(v.__module__, k)))
                               for k, v in inspect.getmembers(paypalrestsdk, inspect.isclass))
    try:
        klass = locate(resource_class_dict[name.lower()])
    except:
        klass = locate(mapping_table[name.lower()])
    return klass
```

Not sure if there is a better place to put the mapping table though.

**Comment by tkrevh** [comment]
Fixed it (tested and fixed also other webhook events, as many didn't have propper mapping).
Here is the PR:
[https://github.com/paypal/PayPal-Python-SDK/pull/218](https://github.com/paypal/PayPal-Python-SDK/pull/218)

**Comment by ClemLivy** [comment]
Nothing new about this?
@tkrevh, your solution doesn't work for me (Python 2.7.13).`klass = locate(mapping_table[name.lower()])` which correspond to `klass = locate('paypalrestsdk.payments.Payout')` return an exception saying 

> ErrorDuringImport: problem in paypalrestsdk.payments.Payout - <type 'exceptions.ImportError'>: No module named Payout

**Comment by carldunham** [comment]
Was this ever even tested? I'm seeing this with sandbox webhook events that include `resource_type` of "invoices", but the resource is under `"resource": {"invoice": {...}}`
